libs/compute/include/hpx/compute/host/block_allocator.hpp

```diff
@@ -35,241 +36,242 @@
-                            },
-                            // cleanup function, called for all elements of
-                            // current partition which succeeded before exception
-                            [p](iterator_type it) { (p + *it)->~U(); });
-                    return std::make_pair(it, last);
-                },
-                // finalize, called once if no error occurred
-                [](std::vector<hpx::future<partition_result_type>>&&) {
-                    // do nothing
-                },
-                // cleanup function, called for each partition which
-                // didn't fail, but only if at least one failed
-                [p](partition_result_type&& r) -> void {
-                    while (r.first != r.second)
-                    {
-                        (p + *r.first)->~U();
-                        ++r.first;
-                    }
-                });
-        }
+        private:
+            policy_type policy_;
+        };
+    }    // namespace detail
+
+    /// The block_allocator allocates blocks of memory evenly divided onto the
+    /// passed vector of targets. This is done by using first touch memory
+    /// placement.
+    ///
+    /// This allocator can be used to write NUMA aware algorithms:
+    ///
+    /// using allocator_type = hpx::compute::host::block_allocator<int>;
+    /// using vector_type = hpx::compute::vector<int, allocator_type>;
+    ///
+    /// auto numa_nodes = hpx::compute::host::numa_domains();
+    /// std::size_t N = 2048;
+    /// vector_type v(N, allocator_type(numa_nodes));
+    ///
+    template <typename T,
+        typename Executor =
+            hpx::parallel::execution::restricted_thread_pool_executor>
+    struct block_allocator
+      : public detail::policy_allocator<T,
+            hpx::parallel::execution::parallel_policy_shim<
+                block_executor<Executor>,
+                typename block_executor<Executor>::executor_parameters_type>>
+    {
+        using executor_type = block_executor<Executor>;
+        using executor_parameters_type =
+            typename executor_type::executor_parameters_type;
+        using policy_type =
+            hpx::parallel::execution::parallel_policy_shim<executor_type,
+                executor_parameters_type>;
+        using base_type = detail::policy_allocator<T, policy_type>;
+        using target_type = std::vector<host::target>;
 
-        // Constructs an object of type T in allocated uninitialized storage
-        // pointed to by p, using placement-new
-        template <typename U, typename... Args>
-        void construct(U* p, Args&&... args)
+        block_allocator()
+          : base_type(policy_type(
+                executor_type(target_type(1)), executor_parameters_type()))
         {
-            executor_.execute(hpx::util::functional::placement_new<U>(), p,
-                std::forward<Args>(args)...);
```

Comment:
I think this function never got instantiated before. Is `execute` some *really* old executor member function? I've changed it to just do placement new directly, but did `execute` have some other behaviour?

Reply:
Yes, this should be `executor_.sync_execute()` instead. That name was changed a long time ago and we missed fixing it here.

Reply:
Well, actually, I think we should use the customization points instead (all over this file). So this would have to use `parallel::execution::sync_execute(executor_, …)`.

Reply:
This was the only that could use a customization point. I've updated it. `bulk_construct` uses a partitioner.